contributor-docs/adrs/adr-021-automated-prop-docs.md

```diff
@@ -0,0 +1,277 @@
+- (optional) A story filed named `{ComponentName}.features.stories.tsx`
+- (optional) A story filed named `{ComponentName}.examples.stories.tsx`
+
+---
+
+If you want a component to be documented, it must be exported from the `.tsx` file. So far this only causes problems for compount components that we export like this: `export const Root = Object.assign(RootComponent, {Item: ItemComponent})`.
+
+This is a problem with `react-typescript-docgen`. TypeDoc will associate subcomponents as "properties" of the exported parent function component.
+
+**Workarounds:**
+- Instead of `export const Root = Object.assign(RootComponent, {Item: ItemComponent})` in the component's `.tsx` file, export each component from the `.tsx` file and then import into the component's `index.ts` file, then export with `Object.assign` from `index.ts`.
+- Export each subcomponent even if you're not actually importing it anywhere else
+
+#### Other challenges
+
+- Prop docs for components like `Button` that accept all relevant HTML props could get _all_ possible props that can be passed to a `<button>` element of just a subset of the most relevant props
+- Generic types are not always written out. For example: `ResponsiveValue<T>` will be rendered instead of `{narrow?: T, regular?: T, wide?: T}`.
+- Still haven't figured out how to skip parsing utility functions (for example: [isNumeric](https://github.com/primer/react/blob/219868e65214d6667c855c22d5c586b2f6b3348d/packages/react/src/utils/isNumeric.tsx)) for inclusion in `components.json`
+- Still haven't figured out how to get Docgen to parse components that are cast using `as PolymorphicForwardRefComponent<E, P>`
+- Still haven't figured out how to get Storybook Docgen data to build the exact same way we build Docgen data for `components.json`
+
+### Comparing third-party tools
+
+I explored the following tools to help us extract component API data:
+
+- [react-docgen-typescript](https://github.com/styleguidist/react-docgen-typescript)
+- [TypeDoc](https://typedoc.org/)
+- [react-docgen](https://react-docgen.dev/)
+- [api-extractor](https://api-extractor.com/)
+
+Alternatively, we could go lower-level and build our own tool from scratch. I have not yet attempted this. See the following examples:
+
+- [Adobe Spectrum DocsTransformer](https://github.com/adobe/react-spectrum/blob/main/packages/dev/parcel-transformer-docs/DocsTransformer.js)
+- [MUI ComponentApiBuilder](https://github.com/mui/material-ui/blob/next/packages/api-docs-builder/ApiBuilders/ComponentApiBuilder.ts)
+
+I found [react-docgen-typescript](https://github.com/styleguidist/react-docgen-typescript) and [TypeDoc](https://typedoc.org/) to be the most viable tools to consider.
+
+Currently, I'd prefer to move forward with [react-docgen-typescript](https://github.com/styleguidist/react-docgen-typescript) because it has given me the best results so far, and it seems possible to customize and improve to better suit our needs.
```

Comment:
What has it looked like customizing it btw? Just was curious 👀 Is it like a plugin system or something else?

Reply:
The primary way to handle this is to use their configuration options: https://github.com/styleguidist/react-docgen-typescript?tab=readme-ov-file#options

However, there's an undocumented option, [shouldIncludeExpression](https://github.com/styleguidist/react-docgen-typescript/blob/master/src/parser.ts#L94), that we could probably use to get more info and go deeper.